extension/encoding/googlecloudlogentryencodingextension/internal/vpcflowlog/parser.go

```diff
@@ -0,0 +1,327 @@
+}
+
+type managedInstanceGroup struct {
+	Name   string `json:"name"`
+	Region string `json:"region"`
+	Zone   string `json:"zone"`
+}
+
+type location struct {
+	ASN       *int64 `json:"asn"`
+	City      string `json:"city"`
+	Continent string `json:"continent"`
+	Country   string `json:"country"`
+	Region    string `json:"region"`
+}
+
+type vpc struct {
+	ProjectID        string `json:"project_id"`
+	SubnetworkName   string `json:"subnetwork_name"`
+	SubnetworkRegion string `json:"subnetwork_region"`
+	VPCName          string `json:"vpc_name"`
+}
+
+type internetRoutingDetails struct {
+	EgressASPath []egressASPath `json:"egress_as_path"`
+}
+
+type egressASPath struct {
+	ASDetails []asDetails `json:"as_details"`
+}
+
+type asDetails struct {
+	ASN *int64 `json:"asn"`
+}
+
+func handleConnection(conn *connection, attr pcommon.Map) {
+	if conn == nil {
+		return
+	}
+
+	// Map protocol number to string
+	if conn.Protocol != nil {
+		if protocolStr, exists := protocolNames[uint32(*conn.Protocol)]; exists {
+			attr.PutStr(string(semconv.NetworkProtocolNameKey), protocolStr)
+		}
+	}
+
+	shared.PutStr(string(semconv.SourceAddressKey), conn.SrcIP, attr)
+	shared.PutStr(string(semconv.DestinationAddressKey), conn.DestIP, attr)
+
+	// Only add port attributes if ports are present
+	shared.PutInt(string(semconv.SourcePortKey), conn.SrcPort, attr)
+	shared.PutInt(string(semconv.DestinationPortKey), conn.DestPort, attr)
+}
+
+func handleNetworkService(ns *networkService, attr pcommon.Map) {
+	if ns == nil {
+		return
+	}
+
+	shared.PutInt(gcpVPCFlowNetworkServiceDSCP, ns.DSCP, attr)
+}
+
+func handleInstance(inst *instance, side flowSide, attr pcommon.Map) error {
+	if inst == nil {
+		return nil
+	}
+
+	// Validate that side is one of our constants
+	if side != src && side != dest {
+		return errors.New("unsupported side passed to handleInstance. Must be one of src or dest")
+	}
+
+	shared.PutStr(fmtAttributeNameUsingSide(gcpVPCFlowInstanceProjectIDTemplate, side), inst.ProjectID, attr)
+	shared.PutStr(fmtAttributeNameUsingSide(gcpVPCFlowInstanceVMRegionTemplate, side), inst.Region, attr)
+	shared.PutStr(fmtAttributeNameUsingSide(gcpVPCFlowInstanceVMNameTemplate, side), inst.VMName, attr)
+	shared.PutStr(fmtAttributeNameUsingSide(gcpVPCFlowInstanceVMZoneTemplate, side), inst.Zone, attr)
+
+	if inst.ManagedInstanceGroup != nil {
+		shared.PutStr(fmtAttributeNameUsingSide(gcpVPCFlowInstanceMIGNameTemplate, side), inst.ManagedInstanceGroup.Name, attr)
+		shared.PutStr(fmtAttributeNameUsingSide(gcpVPCFlowInstanceMIGZoneTemplate, side), inst.ManagedInstanceGroup.Zone, attr)
+	}
+
+	return nil
+}
+
+func handleLocation(loc *location, side flowSide, attr pcommon.Map) error {
+	if loc == nil {
+		return nil
+	}
+
+	// Validate that side is one of our constants
+	if side != src && side != dest {
+		return errors.New("unsupported side passed to handleLocation. Must be one of src or dest")
```

Comment:
```suggestion
		return fmt.Errorf(`unsupported side %q passed to handleLocation. Must be one of %s" , side, validSides)
```

What about being a bit more explicit in this error?

Reply:
Done, and added tests